misc/x509/src/lib.rs

```diff
@@ -0,0 +1,125 @@
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! TLS configuration for `libp2p-quic`.
+#![deny(
+    exceeding_bitshifts,
+    invalid_type_param_default,
+    missing_fragment_specifier,
+    mutable_transmutes,
+    no_mangle_const_items,
+    overflowing_literals,
+    patterns_in_fns_without_body,
+    pub_use_of_private_extern_crate,
+    unknown_crate_types,
+    const_err,
+    order_dependent_trait_objects,
+    illegal_floating_point_literal_pattern,
+    improper_ctypes,
+    late_bound_lifetime_arguments,
+    non_camel_case_types,
+    non_shorthand_field_patterns,
+    non_snake_case,
+    non_upper_case_globals,
+    no_mangle_generic_items,
+    path_statements,
+    private_in_public,
+    safe_packed_borrows,
+    stable_features,
+    type_alias_bounds,
+    tyvar_behind_raw_pointer,
+    unconditional_recursion,
+    unused,
+    unused_allocation,
+    unused_comparisons,
+    unused_mut,
+    unreachable_pub,
+    while_true,
+    anonymous_parameters,
+    bare_trait_objects,
+    elided_lifetimes_in_paths,
+    missing_copy_implementations,
+    missing_debug_implementations,
+    missing_docs,
+    single_use_lifetimes,
+    trivial_casts,
+    trivial_numeric_casts,
+    unused_extern_crates,
+    unused_import_braces,
+    unused_qualifications,
+    clippy::all
+)]
+#![forbid(unsafe_code)]
+
+mod certificate;
+mod verifier;
+
+use std::sync::Arc;
+pub use verifier::extract_peerid_or_panic;
+
+const LIBP2P_SIGNING_PREFIX: [u8; 21] = *b"libp2p-tls-handshake:";
+const LIBP2P_SIGNING_PREFIX_LENGTH: usize = LIBP2P_SIGNING_PREFIX.len();
+const LIBP2P_OID_BYTES: &[u8] = &[43, 6, 1, 4, 1, 131, 162, 90, 1, 1];
+
+fn make_client_config(
+    certificate: rustls::Certificate, key: rustls::PrivateKey,
+    verifier: Arc<verifier::Libp2pCertificateVerifier>,
+) -> rustls::ClientConfig {
+    let mut crypto = rustls::ClientConfig::new();
+    crypto.versions = vec![rustls::ProtocolVersion::TLSv1_3];
+    crypto.alpn_protocols = vec![b"libp2p".to_vec()];
+    crypto.enable_early_data = false;
+    crypto
+        .set_single_client_cert(vec![certificate], key)
+        .expect("we have a valid certificate; qed");
```

Comment:
I do not think unwrapping the result is the right thing to do here. The method can fail because the private key is not one of the expected formats and this may change from one rustls version to another. Better return a `Result` from `make_client_config`. In general I think this crate unwraps way too many results from other crates. If the types express fallible computations asserting the right set of parameters is guaranteed to return an `Ok` result does not help at all. It is not realistic to review all those crates (and their future versions) to check if this holds true or not so I strongly suggest to use your own error type in this crate and return results instead of unwrapping.

Reply:
Fixed.